gh-124153: Implement `PyType_GetBaseByToken()` and `Py_tp_token` slot

neonene opened the pull request:
<!-- gh-issue-number: gh-124153 -->
* Issue: gh-124153
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--124163.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

neonene committed:
Implement the feature

neonene committed:
Use it in _ctypes

neonene committed:
NEWS

neonene committed:
edit a comment

neonene committed:
typo

encukou commented on Objects/typeobject.c:
```diff
@@ -5274,6 +5287,129 @@ _PyType_GetModuleByDef2(PyTypeObject *left, PyTypeObject *right,
+    do { \
+        *result = (PyTypeObject *)(BASE); \
+        return (RET); \
+    } while (0)
```
This macro adds 6 lines to avoid 5 simpler lines. Consider writing the assignments out instead.

encukou reviewed (approved):
Looks great, thank you!
I have just one style suggestion. But if you'd like to keep it we can merge the PR as it is.

neonene committed:
cleanup

neonene commented on Objects/typeobject.c:
```diff
@@ -5274,6 +5287,130 @@ _PyType_GetModuleByDef2(PyTypeObject *left, PyTypeObject *right,
+        return -1;
+    }
+
+    if (!_PyType_HasFeature(type, Py_TPFLAGS_HEAPTYPE)) {
```
When I observed a good performance, the line had a typo (missing a NOT), though `check_base_by_token()` was unused during a PGO test.  The `telco` got slower by 5-10% once I corrected, which was actually fixed by one of the following:

* Use `if-else` statement

* Replace `_PyType_HasFeature()` with the public `PyType_HasFeature()`, not with ineffective `(type->tp_flags & Py_TPFLAGS_HEAPTYPE)`

They are not used in `PyType_GetBaseByToken()`.

neonene committed:
fix possible perf regression

neonene committed:
rely only on if-else statements to fix the perf

This chains the branches in PyType_GetBaseByToken() as well, which is effective when MSVC is in better condition on PGO.

neonene committed:
Merge branch 'main' into tp_token

neonene committed:
remove a comment
